tests/test_views.py

```diff
@@ -3370,6 +3371,89 @@ def test_resend_confirmation(self, send_mail):
 
+    @mock.patch('framework.auth.views.mails.send_mail')
+    def test_click_confirmation_email(self, send_mail):
+        email = 'test@example.com'
+        token = self.user.add_unconfirmed_email(email)
+        self.user.save()
+        self.user.reload()
+        assert_equal(self.user.email_verifications[token]['confirmed'], False)
+        url = '/confirm/{}/{}/?existing_user={}'.format(self.user._id, token, self.user.username)
+        res = self.app.get(url)
+        self.user.reload()
+        assert_equal(self.user.email_verifications[token]['confirmed'], True)
+        assert_equal(res.status_code, 302)
+        login_url = '/login/?existing_user={}'.format(urllib.quote_plus(self.user.username))
+        assert_in(login_url, res.body)
+
+    def test_get_email_to_add_no_email(self):
+        url = api_url_for('verified_email_get')
+        res = self.app.get(url, auth=self.auth)
+        assert_equal(res.json_body, [])
+
+    def test_get_unconfirmed_email(self):
+        email = 'test@example.com'
+        self.user.add_unconfirmed_email(email)
+        self.user.save()
+        self.user.reload()
+        url = api_url_for('verified_email_get')
+        res = self.app.get(url, auth=self.auth)
+        assert_equal(res.json_body, [])
+
+    def test_get_email_to_add(self):
+        email = 'test@example.com'
+        token = self.user.add_unconfirmed_email(email)
+        self.user.save()
+        self.user.reload()
+        assert_equal(self.user.email_verifications[token]['confirmed'], False)
+        url = '/confirm/{}/{}/?existing_user={}'.format(self.user._id, token, self.user.username)
+        self.app.get(url)
+        self.user.reload()
+        assert_equal(self.user.email_verifications[token]['confirmed'], True)
+        url = api_url_for('verified_email_get')
+        res = self.app.get(url, auth=self.auth)
+        assert_equal(res.json_body[0]['address'], 'test@example.com')
+
+    def test_add_email(self):
+        email = 'test@example.com'
+        token = self.user.add_unconfirmed_email(email)
+        self.user.save()
+        self.user.reload()
+        assert_equal(self.user.email_verifications[token]['confirmed'], False)
+        url = '/confirm/{}/{}/?existing_user={}'.format(self.user._id, token, self.user.username)
+        self.app.get(url)
+        self.user.reload()
+        get_email_url = api_url_for('verified_email_get')
+        email_to_add = self.app.get(get_email_url, auth=self.auth)
+        put_email_url = api_url_for('verified_email_add')
+        res = self.app.put_json(put_email_url, email_to_add.json_body[0], auth=self.user.auth)
+        self.user.reload()
+        assert_equal(res.json_body['status'], 'success')
+        assert_equal(self.user.emails[1], 'test@example.com')
+
+    def test_add_email_merge(self):
+        email = "copy@cat.com"
+        dupe = UserFactory(
+            username=email,
+            emails=[email]
+        )
+        dupe.set_password("copycat")
```

Comment:
Is it necessary to set the password in this test? This is minor, but bcrypt-ing passwords can slow down tests significantly.